pandas/_libs/index.pyx

```diff
@@ -606,25 +606,111 @@ cdef class BaseMultiIndexCodesEngine:
                        in zip(self.levels, zip(*target))]
         return self._codes_to_ints(np.array(level_codes, dtype='uint64').T)
 
-    def get_indexer(self, object target, object method=None,
-                    object limit=None):
+    def get_indexer(self, object target) -> np.ndarray:
+        """
```

Comment:
why are you changing this api? can't you rather just call a helper function if you see method / limit?

Reply:
rationale is that in order to do reindexing with filling (not necessary when not filling though), without using "carrying" logic, it's necessary to know the full set of tuples, meaning we have a few options, as far as I can tell:

(1) pass the tuples in from the `MultiIndex` class.  The current implementation uses this; but I separated it for w/ and w/o filling to avoid adding new arguments to `get_indexer()`.

(2) since the MultiIndex's `_engine` knows only of the levels and the int representation of the codes, we could reconstruct the tuples from this, after doing appropriate place-value logic (i.e. inverting `self._codes_to_ints()`, see [1] for more detail if desired), and then proceed with the algorithm in `get_indexer_and_fill()`

(3) have the MultiIndex's engine store a reference to the tuples as well, but I thought a change like that might be inappropriate for this PR.  Happy to do whatever you prefer here!


**footnotes, not necessary to read**:
[1] this is basically the inversion of the `_codes_to_ints()` method in the `MultiIndex` class (source code [here](https://github.com/pandas-dev/pandas/blob/master/pandas/core/indexes/multi.py#L85-L113)).  I'd like to avoid doing this if possible, since it would be a (complex) computation of a value which can just be passed in, but it's definitely mathematically possible, and here's an example of how to do it, if this is unclear.

given:
```
>>> mi = pd.MultiIndex.from_tuples([
...     (1, 3),
...     (2, 1),
...     (2, 2),
...     (2, 4),
...     (5, 0)
... ])
>>> mi
MultiIndex([(1, 3),
            (2, 1),
            (2, 2),
            (2, 4),
            (5, 0)],
           )
```
we have the following info available to us:
```
>>> mi._engine.levels
FrozenList([[1, 2, 5], [0, 1, 2, 3, 4]])
>>> mi._engine.vgetter()
array([12, 18, 19, 21, 25], dtype=uint64)
```
knowing this, we can reconstruct that we need ceil(log_2(3)) = 2 bits for the 0-level values, ceil(log_2(5)) = 3 bits for the 1-level values, and so we could determine that:
- 12 = 1(2^3) + 4(2^0)
- 18 = 2(2^3) + 2(2^0)
- 19 = 2(2^3) + 3(2^0)
- 21 = 2(2^3) + 5(2^0)
- 25 = 3(2^3) + 1(2^0)

and so, after decrementing all of those coefficients by 1 to account for 0-indexing, this would give us the following (per-level) indexer-like list of indexes:
- (0, 3)
- (1, 1)
- (1, 2)
- (1, 4)
- (2, 0)
and if we get the values of `mi._engine.levels` at these (per-level) indices, we get:
- (1, 3)
- (2, 1)
- (2, 2)
- (2, 4)
- (5, 0)
i.e. the MultiIndex's list of tuples